proposals/NNNN-add-customdebugdescription-conformance-to-anykeypath.md

```diff
@@ -0,0 +1,158 @@
+
+### Dealing with missing data
+
+There are two known cases where data might not be available: 
+
+1. type metadata has not been emitted because the target was built using the `swift-disable-reflection-metadata` flag
+2. the linker has stripped the symbol names we're trying to look up
+
+In these cases, we would print the following: 
+
+#### Offset case
+`<offset [x] ([typename])>` where `x` is the memory offset we read from the reflection metadata, and `typename` is the type that will be returned. 
+So 
+```
+print(\Theme.backgroundColor) // outputs "\Theme.<offset 0 (Color)>"
+```
+
+#### `lookupSymbol` failure case
+
+In this case we'll print the address-in-memory as hex, plus the type name: 
+```
+print(\Theme.overlay) // outputs \Theme.<computed 0xABCDEFG (Color)>
+```
+
+As it might be difficult to correlate a memory address with the name of the function, the type name may be useful here to provide extra context. 
+
+## Source compatibility
+
+Programs that extend `AnyKeyPath` to implement `CustomDebugStringConvertible` themselves will no longer compile and the authors of such code will have to delete the conformance. Based on a search of Github, there are currently no publically available Swift projects that do this. 
+
+Calling `print` on a KeyPath will, of course, produce different results than before. 
+
+It is unlikely that any existing Swift program is depending on the existing behavior in a production context. While it is likely that someone, somewhere has written code in unit tests that depends on the output of this function, any issues that result will be easy for the authors of such code to identify and fix, and will likely represent an improvement in the readability of those tests. 
+
+## Effect on ABI stability
+
+This proposal will add a new var & protcol confromance to the Standard Library's ABI. It will be availability guarded appropriately. 
+
+## Effect on API resilience
+
+The implementation of `debugDescription` might change after the inital work to implement this proposal is done. In particular, the output format will not be guarenteed to be stable. Here are a few different changes we might anticipate making: 
+
+- As new features are added to the compiler, there may be new metadata available in the binary to draw from. One example would be lookup tables of KeyPath segment to human-readable-name or some other unique, stable identifier
+- Whenever a new feature is added to `KeyPath`, it will need to be reflected in the output of this function. For example, the `KeyPath`s produced by [\_forEachFieldWithKeyPath](https://github.com/apple/swift/blob/main/stdlib/public/core/ReflectionMirror.swift#L324) are incomplete, in the sense that they merely set a value at in offset in memory and do not call `didSet` observers. If this function were ever publically exposed, it would be useful if this was surfaced in the debugging information. 
```

Comment:
* use permalink?
* "publically" → "publicly"

```suggestion
- Whenever a new feature is added to `KeyPath`, it will need to be reflected in the output of this function. For example, the `KeyPath`s produced by [\_forEachFieldWithKeyPath](https://github.com/apple/swift/blob/main/stdlib/public/core/ReflectionMirror.swift#L324) are incomplete, in the sense that they merely set a value at in offset in memory and do not call `didSet` observers. If this function were ever publicly exposed, it would be useful if this was surfaced in the debugging information. 
```